### PR TITLE
MGMT-5444: Set 24 hours timeout for installation to complete

### PR DESCRIPTION
**Pull request by eranco74**
# Assisted Pull Request

## Description

MGMT-5444: Set 24 hours timeout for installation to complete
    
The installation might hang while the assisted service wait for user action
e.g. host with wrong boot order or a host that didn't return after reboot.
Since some certificates generated by the installer are valid for 24 hours
it's unlikely for the installation to complete successfully once 24 hours pass.
In order to avoid debugging strange installation failures due to TLS
certificates expiration, this code adds a global timeout of 24 hours for the installation.
This timeout is calculated according to the `cluster.InstallStartedAt`
which get updated in PostPrepareForInstallation

- Should this PR be tested by the reviewer? no
- Is this PR relying on CI for an e2e test run? yes, for regression 
- Should this PR be tested in a specific environment? no
- Any logs, screenshots, etc that can help with the review process? no

-->

## List all the issues related to this PR
https://issues.redhat.com/browse/MGMT-5444
- [ ] New Feature
- [x] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [x] Automation (CI, tools, etc)
- [x] Cloud
- [x] Operator Managed Deployments
- [ ] None

## How was this code tested?

- [x] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [ x] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [x] No tests needed

## Assignees

/cc @tsorya
/cc @osherdp 

## Checklist

- [x] Title and description added to both, commit and PR.
- [x] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [x] Reviewers have been listed
- [x] This change does not require a documentation update (docstring, `docs`, README, etc)
- [x] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- [ ] Are the title and description (in both PR and commit) meaningful and clear?
- [ ] Is there a bug required (and linked) for this change?
- [ ] Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: internal/cluster/common.go [hunk 1]
```diff
@@ -35,6 +35,7 @@ const (
 	statusInfoPreparingForInstallationTimeout = "Preparing cluster for installation timeout"
 	statusInfoPendingForInput                 = "User input required"
 	statusInfoError                           = "cluster has hosts in error"
+	statusInfoTimeout                         = "cluster installation timed out while pending user action"
```
Comment (osherdp) [inline internal/cluster/common.go:38]: a really slow installation cannot fail on it as well?
or some other step will timeout before the general timeout

File: internal/cluster/transition.go [hunk 2]
```diff
@@ -409,14 +410,27 @@ func (th *transitionHandler) enoughMastersAndWorkers(sCluster *stateCluster, sta
 	return false
 }
 
+//check if installation reach to timeout
+func (th *transitionHandler) IsInstallationTimedOut(sw stateswitch.StateSwitch, args stateswitch.TransitionArgs) (bool, error) {
+	sCluster, ok := sw.(*stateCluster)
+	if !ok {
+		return false, errors.New("IsInstallationTimedOut incompatible type of StateSwitch")
+	}
+	// can happen if the service was rebooted or somehow the async part crashed.
```
Comment (osherdp) [inline internal/cluster/transition.go:419]: the service was rebooted?

File: internal/cluster/transition.go [hunk 3]
```diff
@@ -409,14 +410,27 @@ func (th *transitionHandler) enoughMastersAndWorkers(sCluster *stateCluster, sta
...
+	if time.Since(time.Time(sCluster.cluster.InstallStartedAt)) > th.installationTimeout {
```
Comment (osherdp) [inline internal/cluster/transition.go:420]: seems like ``InstallStartedAt`` is already ``time.Time``
can we remove the ``time.Time(...)``?

File: internal/cluster/statemachine.go [hunk 4]
```diff
@@ -193,6 +193,17 @@ func NewClusterStateMachine(th *transitionHandler) stateswitch.StateMachine {
 		PostTransition:   th.PostRefreshCluster(statusInfoError),
 	})
 
+	// Installation timeout while pending user action
```
Comment (osherdp) [inline internal/cluster/statemachine.go:196]:
```suggestion
	// Installation timeout while pending user action (a manual booting from installation disk)
```

**Review by osherdp — CHANGES_REQUESTED** [review]

File: internal/cluster/transition.go (hunk 3 above)
Comment (eranco74) [reply internal/cluster/transition.go:420]: it's `strfmt.DateTime`

File: internal/cluster/transition.go (hunk 2 above)
Comment (eranco74) [reply internal/cluster/transition.go:419]: ooops

File: internal/cluster/statemachine.go (hunk 4 above)
Comment (eranco74) [reply internal/cluster/statemachine.go:196]: Updated.
@avishayt any preferences here?

File: internal/cluster/common.go (hunk 1 above)
Comment (eranco74) [reply internal/cluster/common.go]: Other steps will timeout and fail the installation.
TMK pending user action is the only status that allows the installation to hang without a timeout.

File: internal/cluster/transition.go (hunk 3 above)
Comment (osherdp) [reply internal/cluster/transition.go:419]: I think it's a reference to ``time.Time``
Anyway, not critical

**Review by osherdp — APPROVED** [review]
looks good to me
unhold whenever you want
/hold

**Comment by avishayt** [comment]
I thought we wanted to time out at 24 hours regardless of the state, but here we're timing out only for pending-user-action.  I understand that it's the most likely case, but why not implement a global timeout?

**Comment by eranco74** [comment]
> I thought we wanted to time out at 24 hours regardless of the state, but here we're timing out only for pending-user-action. I understand that it's the most likely case, but why not implement a global timeout?

The 24 hours timeout is relevant only during installation.
TTBOMK, `Installing` and `pending user action` are the only relevant cases.
Installing timeout is covered by the stages timeouts.

Let's start with this one and see that it doesn't cause false failures.

**Comment by eranco74** [comment]
This is how it looks in the assisted-service logs:

time="2021-08-09T15:55:14Z" level=info msg="cluster 13111fd3-e2cc-4a60-ac40-99779d2e4810 has been updated with the following updates [status error status_info cluster installation timed out while pending user action (a manual booting from installation disk) status_updated_at 2021-08-09T15:55:14.673Z install_complet>

**Pull request merged**